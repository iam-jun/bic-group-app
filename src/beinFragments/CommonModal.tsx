import React, {useRef} from 'react';

import BottomSheet from '~/beinComponents/BottomSheet';
import {useKeySelector} from '~/hooks/selector';
import modalKeySelector from '~/store/modal/keySelector';
import {useDispatch} from 'react-redux';
import modalActions from '~/store/modal/actions';

const CommonModal = () => {
  const modalizeRef = useRef();

  const dispatch = useDispatch();

  const modal = useKeySelector(modalKeySelector.modal);
  const {isOpen, ContentComponent} = modal || {};

  const _onClose = () => {
    dispatch(modalActions.hideModal());
  };

  return (
    <BottomSheet
      modalizeRef={modalizeRef}
      isContextMenu={false}
      isOpen={isOpen}
      onClose={_onClose}
      ContentComponent={ContentComponent}
    />
  );
};

export default CommonModal;
