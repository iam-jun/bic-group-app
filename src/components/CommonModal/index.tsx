import React, { useEffect, useRef } from 'react';
import {
  Modal, Platform, StyleSheet, TouchableOpacity, View,
} from 'react-native';
import { useDispatch } from 'react-redux';
import { ExtendedTheme, useTheme } from '@react-navigation/native';
import { useSafeAreaInsets } from 'react-native-safe-area-context';
import BottomSheet from '~/baseComponents/BottomSheet';
import { useKeySelector } from '~/hooks/selector';
import modalActions from '~/storeRedux/modal/actions';
import modalKeySelector from '~/storeRedux/modal/keySelector';
import spacing from '~/theme/spacing';
import Button from '~/beinComponents/Button';
import Icon from '~/beinComponents/Icon';
import Text from '~/beinComponents/Text';

const CommonModal = () => {
  const modalizeRef = useRef<any>();

  const dispatch = useDispatch();
  const theme: ExtendedTheme = useTheme()
  const styles = themeStyles(theme);

  const modal = useKeySelector(modalKeySelector.modal);
  const {
    isOpen,
    isFullScreen,
    titleFullScreen,
    ContentComponent,
    props,
    useAppBottomSheet = true,
    closeOutSide = true,
  } = modal || {};

  useEffect(
    () => {
      if (!isOpen) {
        modalizeRef?.current?.close?.();
      }
    }, [isOpen],
  );

  const _onClose = () => {
    closeOutSide && dispatch(modalActions.hideModal());
  };

  if (isFullScreen) {
    return (
      <Modal
        visible={isOpen}
        transparent
        animationType="slide"
        onRequestClose={() => dispatch(modalActions.hideModal())}
      >
        <View testID="common_modal.center" style={styles.fullScreenContainer}>
          <View style={styles.fullScreenHeader}>
            <Text.H4 style={styles.titleFullScreen} numberOfLines={2}>
              {titleFullScreen}
            </Text.H4>
            <Button style={styles.btnClose} onPress={_onClose}>
              <Icon icon="iconCloseSmall" />
            </Button>
          </View>
          <View style={{ flex: 1 }}>
            {ContentComponent}
          </View>
        </View>
      </Modal>
    );
  }

  if (!useAppBottomSheet) {
    return (
      <Modal
        visible={isOpen}
        transparent
        animationType="slide"
        onRequestClose={() => {
          dispatch(modalActions.hideModal());
        }}
      >
        <TouchableOpacity
          testID="common_modal.center"
          activeOpacity={1}
          style={styles.appModalContainer}
          onPress={_onClose}
        >
          {ContentComponent}
        </TouchableOpacity>
      </Modal>
    );
  }

  return (
    <BottomSheet
      modalizeRef={modalizeRef}
      isOpen={isOpen}
      onClose={_onClose}
      ContentComponent={ContentComponent}
      {...props}
    />
  );
};

const themeStyles = (theme: ExtendedTheme) => {
  const { colors } = theme;
  const insets = useSafeAreaInsets();

  return StyleSheet.create({
    appModalContainer: {
      flex: 1,
      backgroundColor: 'rgba(12, 13, 14, 0.5)',
      justifyContent: 'center',
    },
    fullScreenContainer: {
      flex: 1,
      backgroundColor: colors.neutral,
      justifyContent: 'center',
      paddingHorizontal: spacing.padding.small,
      paddingTop: Platform.OS === 'android' ? 0 : insets.top,
      paddingBottom: insets.bottom + spacing.padding.small,
    },
    fullScreenHeader: {
      flexDirection: 'row',
      paddingVertical: spacing.padding.small,
      alignItems: 'center',
    },
    titleFullScreen: {
      flex: 1,
      marginLeft: spacing.margin.large,
      paddingVertical: spacing.padding.small,
    },
    btnClose: { paddingHorizontal: spacing.padding.extraLarge },
  });
}

export default CommonModal;
