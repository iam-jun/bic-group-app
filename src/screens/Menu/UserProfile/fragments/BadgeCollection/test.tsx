import React from 'react';
import {
  View, Dimensions, FlatList, StyleSheet,
} from 'react-native';
import { ExtendedTheme, useTheme } from '@react-navigation/native';
import Avatar from '~/baseComponents/Avatar';
import Icon from '~/baseComponents/Icon';
import spacing, { borderRadius } from '~/theme/spacing';

const Grid = ({ data }: {data: any[]}) => {
  const deviceWidth = Dimensions.get('window').width;

  const itemWidth = 48; // Constant item width
  const numColumns = Math.floor((deviceWidth - 10 * 2) / (itemWidth + 6 * 2));
  const rowWidth = numColumns * (itemWidth + 6 * 2);
  const paddingLeft = (deviceWidth - (10 * 2) - rowWidth) / 2;

  const theme: ExtendedTheme = useTheme();
  const { colors } = theme;
  const styles = themeStyles(theme);

  const isSelected = false;

  const renderGridItem = ({ item, index }: any) =>
    // Render each grid item here
    (
      <View
        key={index}
        style={{
          margin: 6,
          borderWidth: isSelected ? 2 : 1,
          borderColor: isSelected ? colors.purple50 : colors.neutral5,
          borderRadius: borderRadius.pill,
        }}
      >
        <Avatar.Medium
          isRounded
          source={{ uri: item?.iconUrl }}
        />
        {isSelected
          ? (
            <View style={styles.iconChose}>
              <Icon size={10} icon="Check" tintColor={colors.white} />
            </View>
          )
          : null}
      </View>
    );

  return (
    <View style={{
      flex: 1,
      justifyContent: 'center',
      alignItems: 'center',
      paddingHorizontal: 10,
      paddingBottom: 10,
    }}
    >

      <View style={{
        // paddingLeft,
      }}
      >
        <FlatList
          data={data}
          renderItem={renderGridItem}
          keyExtractor={(item, index) => index.toString()}
          numColumns={numColumns}
        />
      </View>
    </View>
  );
};

const themeStyles = (theme: ExtendedTheme) => {
  const { colors } = theme;

  return StyleSheet.create({
    container: {
      flex: 1,
      paddingTop: spacing.padding.large,
    },
    iconChose: {
      backgroundColor: colors.purple50,
      borderRadius: borderRadius.pill,
      height: 20,
      width: 20,
      alignItems: 'center',
      justifyContent: 'center',
      position: 'absolute',
      zIndex: 1,
      right: 0,
    },
  });
};

export default Grid;
