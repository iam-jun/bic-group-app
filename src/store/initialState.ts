import {appInitState} from '~/store/app/reducer';
import {modalInitState} from '~/store/modal/reducer';
import {authInitState} from '~/screens/Auth/redux/reducer';
import {postInitState} from '~/screens/Post/redux/reducer';
import {groupInitState} from '~/screens/Groups/redux/reducer';
import {homeInitState} from '~/screens/Home/redux/reducer';
import {notiInitState} from '~/screens/Notification/redux/reducer';
import {menuInitState} from '~/screens/Menu/redux/reducer';
import {noInternetInitState} from '~/screens/NoInternet/redux/reducer';
import {mentionInputInitState} from '~/beinComponents/inputs/MentionInput/redux/reducer';

const initialState = {
  app: appInitState,
  modal: modalInitState,
  auth: authInitState,
  post: postInitState,
  groups: groupInitState,
  home: homeInitState,
  notifications: notiInitState,
  menu: menuInitState,
  noInternet: noInternetInitState,
  mentionInput: mentionInputInitState,
};

export default initialState;
