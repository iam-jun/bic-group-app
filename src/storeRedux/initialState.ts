import { appInitState } from '~/storeRedux/app/reducer';
import { modalInitState } from '~/storeRedux/modal/reducer';
import { authInitState } from '~/storeRedux/auth/reducer';
import { postInitState } from '~/storeRedux/post/reducer';
import { groupInitState } from '~/storeRedux/groups/reducer';
import { homeInitState } from '~/storeRedux/home/reducer';
import { notiInitState } from '~/storeRedux/notification/reducer';
import { menuInitState } from '~/storeRedux/menu/reducer';
import { noInternetInitState } from '~/storeRedux/network/reducer';
import { mentionInputInitState } from '~/beinComponents/inputs/MentionInput/redux/reducer';
import { giphyInitState } from './giphy/reducer';

const initialState = {
  app: appInitState,
  modal: modalInitState,
  auth: authInitState,
  post: postInitState,
  groups: groupInitState,
  home: homeInitState,
  notifications: notiInitState,
  menu: menuInitState,
  noInternet: noInternetInitState,
  mentionInput: mentionInputInitState,
  giphy: giphyInitState,
};

export default initialState;
