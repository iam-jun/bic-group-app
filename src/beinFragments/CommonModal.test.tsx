import * as React from 'react';
import {View} from 'react-native';
import {cleanup, waitFor} from '@testing-library/react-native';

import {fireEvent, renderWithRedux, configureStore} from '~/test/testUtils';
import CommonModal from './CommonModal';
import initialState from '~/store/initialState';

afterEach(cleanup);

describe('CommonModal component', () => {
  const mockStore = configureStore([]);

  const fake_data = {
    isOpen: true,
    ContentComponent: <View style={{height: 100, backgroundColor: 'blue'}} />,
    props: {
      modalStyle: {borderTopLeftRadius: 20, borderTopRightRadius: 20},
    },
  };

  it(`renders correctly`, () => {
    const storeData = {...initialState};
    // @ts-ignore
    storeData.modal.modal = fake_data;
    const store = mockStore(storeData);
    const rendered = renderWithRedux(<CommonModal />, store);
    expect(rendered.toJSON()).toMatchSnapshot();
  });

  it(`renders correctly with useAppBottomSheet = false`, () => {
    const storeData = {...initialState};
    // @ts-ignore
    storeData.modal.modal = {...fake_data, useAppBottomSheet: false};
    const store = mockStore(storeData);
    const rendered = renderWithRedux(<CommonModal />, store);

    expect(rendered.toJSON()).toMatchSnapshot();

    const customModalComponent = rendered.getByTestId('common_modal.center');
    expect(customModalComponent).toBeDefined();
  });

  it(`renders correctly style with appModalStyle`, () => {
    const storeData = {...initialState};
    // @ts-ignore
    storeData.modal.modal = {
      ...fake_data,
      useAppBottomSheet: false,
      appModalStyle: {height: 200},
    };
    const store = mockStore(storeData);
    const rendered = renderWithRedux(<CommonModal />, store);

    expect(rendered.toJSON()).toMatchSnapshot();

    const customModalComponent = rendered.getByTestId('common_modal.center');
    expect(customModalComponent).toBeDefined();
    expect(customModalComponent.props.style?.height).toBe(200);
  });
});
