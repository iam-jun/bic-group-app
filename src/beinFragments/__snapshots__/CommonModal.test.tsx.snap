// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommonModal component renders correctly 1`] = `
<View>
  <View>
    <View
      style={
        Object {
          "backgroundColor": "blue",
          "height": 100,
        }
      }
    />
  </View>
</View>
`;

exports[`CommonModal component renders correctly with useAppBottomSheet = false 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  transparent={true}
  visible={true}
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "rgba(12, 13, 14, 0.5)",
        "flex": 1,
        "justifyContent": "center",
        "opacity": 1,
      }
    }
    testID="common_modal.center"
  >
    <View
      style={
        Object {
          "backgroundColor": "blue",
          "height": 100,
        }
      }
    />
  </View>
</Modal>
`;
